Only store unique store and forward messages

neonknight64 opened the pull request:
## Description
- These changes allow only unique messages to be stored by the store and forward system.
- This is achieved by comparing the body hash of a new message to the hashes of the currently stored set of messages. Only if the body content is unique is the message stored by the store and forward database.

## Motivation and Context
Limit unnecessary storage of messages with duplicate content.

## How Has This Been Tested?
Modified the store and forward database insert_message test to also test the behaviour when a duplicate message is inserted.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

sdbondi commented on comms/dht/migrations/2020-05-14-103121_add_body_hash/up.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE stored_messages
+    ADD body_hash TEXT;
```
We can let the database do the unique checking by using a UNIQUE INDEX. 

Can add after adding the column:
`CREATE UNIQUE INDEX uidx_stored_messages_body_hash ON stored_messages (body_hash);`

If a duplicate hash is inserted, an error will be thrown. You'd need to handle that error case in some way.Perhaps we should make the insert_message function idempotent i.e. if the message hash is there already (duplicate key error) , return `Ok(())`, not sure if we want to call the function `insert_message_if_unique` or something - You'll need to check into the details of how diesel handles unique indexes and insert errors (https://docs.diesel.rs/diesel/result/enum.DatabaseErrorKind.html). 

Alternatively, you can add StorageError::UniqueViolation and handle that one level up, whichever results in clearer behaviour and code.

neonknight64 replied:
Change applied, the database will perform the unique checking and the `insert_message_if_unique` function was made idempotent.

sdbondi commented on comms/dht/src/store_forward/database/mod.rs:
```diff
@@ -48,10 +48,18 @@ impl StoreAndForwardDatabase {
+                    .get_results::<i32>(conn)?;
+
+                if duplicate_msg_ids.is_empty() {
+                    diesel::insert_into(stored_messages::table)
+                        .values(message)
+                        .execute(conn)?;
```
Handle a `UniqueViolation` in some way and remove the extra query above.

neonknight64 replied:
Extra query was removed and the `UniqueViolation` error was handled.

sdbondi reviewed (commented):
Awesome! Just a suggestion to use the database engine to do unique checking.

sdbondi commented on comms/dht/migrations/2020-05-14-103121_add_body_hash/down.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE stored_messages
+    DROP COLUMN body_hash;
```
Need to add the counterpart to `CREATE UNIQUE INDEX` (i.e. DROP INDEX) above ALTER TABLE, otherwise I believe it'll error(not sure) on the down migration

neonknight64 replied:
Change applied.

sdbondi commented on comms/dht/migrations/2020-05-14-103121_add_body_hash/up.sql:
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE stored_messages
+    ADD body_hash TEXT;
+
+CREATE UNIQUE INDEX uidx_stored_messages_body_hash ON stored_messages (body_hash);
```
Tested migration and works well (nulls are not counted towards the unique index as expected)

neonknight64 replied:
That is great!

sdbondi reviewed (commented):
Thanks - tested on my base node and this works - just need to complete the down migration

Commit:
- Limit unnecessary storage of messages with duplicate content.
- These changes allow only unique messages to be stored by the store and forward system.
- This is achieved by comparing the body hash of a new message to the hashes of the currently stored set of messages. Only if the body content is unique is the message stored by the store and forward database.
- Modified the store and forward database insert_message test to also test the behaviour when a duplicate message is inserted.

sdbondi reviewed (approved):
:+1:
